### PR TITLE
fix(object_store): fs and hdfs object store should set atomic_write_dir

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
as title, all operation will first apply do `root/atomic_write_dir` then apply to `root/data_directory`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/object_store/src/object/opendal_engine/fs.rs [hunk 1]
```diff
@@ -25,6 +25,7 @@ impl OpendalObjectStore {
         let mut builder = Fs::default();
 
         builder.root(&root);
+        builder.atomic_write_dir(&root);
```
Comment (BugenZhao) [inline src/object_store/src/object/opendal_engine/fs.rs:28]: I don't think passing the same directory is correct. 😅 Please test the PR before rushing it out.

Reference: https://github.com/apache/opendal/pull/991

File: src/object_store/src/object/opendal_engine/hdfs.rs [hunk 2]
```diff
@@ -26,7 +26,7 @@ impl OpendalObjectStore {
         // Set the name node for hdfs.
         builder.name_node(&namenode);
         builder.root(&root);
-
+        builder.atomic_write_dir(&root);
```
Comment (BugenZhao) [inline src/object_store/src/object/opendal_engine/hdfs.rs:29]: What about WebHDFS?

Comment (wcy-fdu) [reply src/object_store/src/object/opendal_engine/hdfs.rs:29]: adding

File: src/object_store/src/object/opendal_engine/fs.rs (hunk 1 above)
Comment (Li0k) [reply src/object_store/src/object/opendal_engine/fs.rs:28]: > I don't think passing the same directory is correct. 😅 Please test the PR before rushing it out.
> 
> Reference: [apache/opendal#991](https://github.com/apache/opendal/pull/991)

+1

Comment (wcy-fdu) [reply src/object_store/src/object/opendal_engine/fs.rs:28]: I think we can build a independent path for applying this setting when creating builder, just like [cluster_id](https://github.com/risingwavelabs/risingwave/blob/4197ad50c766c8d086d29948ef34e13582e72251/src/meta/src/hummock/manager/mod.rs#L3351),  and make sure this directory is not used.

File: src/object_store/src/object/opendal_engine/fs.rs [hunk 3]
```diff
@@ -17,14 +17,18 @@ use opendal::services::Fs;
...
 use crate::object::ObjectResult;
+
 impl OpendalObjectStore {
     /// create opendal fs engine.
     pub fn new_fs_engine(root: String) -> ObjectResult<Self> {
         // Create fs backend builder.
         let mut builder = Fs::default();
 
         builder.root(&root);
+        let atomic_write_dir = format!("{}{}", root, ATOMIC_WRITE_DIR);
```
Comment (BugenZhao) [inline src/object_store/src/object/opendal_engine/fs.rs:30]: Do we guarantee that `root` always ends with the separator (`/`)?

Comment (wcy-fdu) [reply src/object_store/src/object/opendal_engine/fs.rs:30]: Yes, root always ends with `/`, and let's just set atomic_write_dir `data_directory/atomic_write_dir`, thus will not  affect the data under `root`. (Actually the `root` is always `/` now).

Comment (hzxa21) [reply src/object_store/src/object/opendal_engine/fs.rs]: I have two questions:
1. Will the temporary files in `data_directory/atomic_write_dir` be deleted if write fails? We may need to check OpenDAL implementation to confirm.
2. Will files in `data_directory/atomic_write_dir` eligible for full gc? cc @zwang28. If these files have `.data` suffix, then I think the answer is yes (See [here](https://github.com/risingwavelabs/risingwave/blob/2e614b9f21494903fd03ec454c13d85418870022/src/storage/src/hummock/sstable_store.rs#L666)). If that is the case, we should blacklist `atomic_write_dir` for full gc.

Comment (hzxa21) [reply src/object_store/src/object/opendal_engine/fs.rs]: > 1. no

oh.... Then maybe it makes sense to let full gc take care of the files in `atomic_write_dir`. No changes are needed.

Comment (Li0k) [reply src/object_store/src/object/opendal_engine/fs.rs]: > I have two questions:
> 
> 1. Will the temporary files in `data_directory/atomic_write_dir` be deleted if write fails? We may need to check OpenDAL implementation to confirm.
> 2. Will files in `data_directory/atomic_write_dir` eligible for full gc? cc @zwang28. If these files have `.data` suffix, then I think the answer is yes (See [here](https://github.com/risingwavelabs/risingwave/blob/2e614b9f21494903fd03ec454c13d85418870022/src/storage/src/hummock/sstable_store.rs#L666)). If that is the case, we should blacklist `atomic_write_dir` for full gc.

I'm also assisting in this investigation.

Comment (hzxa21) [reply src/object_store/src/object/opendal_engine/fs.rs]: > > I have two questions:
> > 
> > 1. Will the temporary files in `data_directory/atomic_write_dir` be deleted if write fails? We may need to check OpenDAL implementation to confirm.
> > 2. Will files in `data_directory/atomic_write_dir` eligible for full gc? cc @zwang28. If these files have `.data` suffix, then I think the answer is yes (See [here](https://github.com/risingwavelabs/risingwave/blob/2e614b9f21494903fd03ec454c13d85418870022/src/storage/src/hummock/sstable_store.rs#L666)). If that is the case, we should blacklist `atomic_write_dir` for full gc.
> 
> I'm also assisting in this investigation.

What do you mean?

**Review by hzxa21 — APPROVED** [review]
LGTM

**Pull request merged**